genericjpa/db/src/main/java/org/hibernate/search/genericjpa/db/events/index/impl/IndexUpdater.java

```diff
@@ -0,0 +1,320 @@
+		this.indexWrapper = new DefaultIndexWrapper( searchIntegrator );
+	}
+
+	public void updateEvent(List<UpdateEventInfo> updateInfos) {
+		if(updateInfos.size() == 0) {
+			return;
+		}
+		this.updateEvent( updateInfos, this.entityProvider );
+	}
+
+	public void updateEvent(List<UpdateEventInfo> updateInfos, EntityProvider provider) {
+		//this is a hack so we can start/end our transactions properly in JTA
+		//as transactions are bound to threads
+		final SearchException[] exception = {null};
+		final CountDownLatch latch = new CountDownLatch( 1 );
+		this.exec.submit(
+				() ->
+				{
+					try {
+						if ( provider instanceof ReusableEntityProvider ) {
+							((ReusableEntityProvider) provider).open();
+						}
+						try {
+							Transaction tx = new Transaction();
+							try {
+								for ( UpdateEventInfo updateInfo : updateInfos ) {
+									Class<?> entityClass = updateInfo.getEntityClass();
+									Map<String, Object> hints = Collections.unmodifiableMap( updateInfo.getHints() );
+									List<Class<?>> inIndexOf = IndexUpdater.this.containedInIndexOf.get( entityClass );
+									if ( inIndexOf != null && inIndexOf.size() != 0 ) {
+										int eventType = updateInfo.getEventType();
+										Object id = updateInfo.getId();
+										switch ( eventType ) {
+											case EventType.INSERT: {
+												Object obj = provider.get( entityClass, id, hints );
+												if ( obj != null ) {
+													IndexUpdater.this.indexWrapper.index( obj, tx );
+												}
+												break;
+											}
+											case EventType.UPDATE: {
+												Object obj = provider.get( entityClass, id, hints );
+												if ( obj != null ) {
+													IndexUpdater.this.indexWrapper.update( obj, tx );
+												}
+												break;
+											}
+											case EventType.DELETE: {
+												IndexUpdater.this.indexWrapper.delete(
+														entityClass, inIndexOf, id, this.entityProvider,
+														tx
+												);
+												break;
+											}
+											default: {
+												LOGGER.warning( "unknown eventType-id found: " + eventType );
+											}
+										}
+									}
+									else {
+										LOGGER.warning( "class: " + entityClass + " not found in any index!" );
+									}
+								}
+								tx.commit();
+							}
+							catch (Exception e) {
+								tx.rollback();
+								LOGGER.log(
+										Level.WARNING,
+										"Error while updating the index! Your index might be corrupt!",
+										e
+								);
+								exception[0] = new SearchException(
+										"Error while updating the index! Your index might be corrupt!",
+										e
+								);
+							}
+						}
+						finally {
+							if ( provider instanceof ReusableEntityProvider ) {
+								((ReusableEntityProvider) provider).close();
+							}
+						}
+					}
+					finally {
+						latch.countDown();
+					}
+
+				}
+		);
+		try {
+			latch.await();
+			//while we did things on a different thread we still
```

Comment:
Also notice that with this pattern you're essentially introducing a bottleneck: all threads coming here which have some update to write will have to delegate to the single thread of your executor, and then wait for it to complete the task. That's a very significant throughput issue.